api/client/commands.go

```diff
@@ -1948,14 +1948,49 @@ func (cli *DockerCli) CmdSearch(args ...string) error {
 
 	utils.ParseFlags(cmd, args, true)
 
+	name := cmd.Arg(0)
 	v := url.Values{}
-	v.Set("term", cmd.Arg(0))
+	v.Set("term", name)
+
+	// Resolve the Repository name from fqn to hostname + name
+	taglessRemote, _ := parsers.ParseRepositoryTag(name)
+	repoInfo, err := registry.ParseRepositoryInfo(taglessRemote)
+	if err != nil {
+		return err
+	}
+
+	cli.LoadConfigFile()
+
+	// Resolve the Auth config relevant for this server
+	authConfig := cli.configFile.ResolveAuthConfig(repoInfo.Index)
+
+	search := func(authConfig registry.AuthConfig) ([]byte, int, error) {
+		buf, err := json.Marshal(authConfig)
+		if err != nil {
+			return nil, -1, err
+		}
+		registryAuthHeader := []string{
+			base64.URLEncoding.EncodeToString(buf),
+		}
 
-	body, _, err := readBody(cli.call("GET", "/images/search?"+v.Encode(), nil, true))
+		return readBody(cli.call("GET", "/images/search?"+v.Encode(), nil, map[string][]string{
+			"X-Registry-Auth": registryAuthHeader,
+		}))
+	}
 
+	body, _, err := search(authConfig)
+	if err != nil && strings.Contains(err.Error(), "status code 401") {
+		fmt.Fprintln(cli.out, "\nPlease login prior to search:")
+		if err = cli.CmdLogin(repoInfo.Index.GetAuthConfigKey()); err != nil {
+			return err
+		}
+		authConfig = cli.configFile.ResolveAuthConfig(repoInfo.Index)
+		body, _, err = search(authConfig)
+	}
```

Comment:
Right, I mimicked how those commands attempt login on first 401.  I am also not happy with how (afaik) there is no access to the status code itself (401 in this case), causing us to resort to parsing error strings.

I could make a pass at cleaning this up, but it would make for a larger PR.  Should I take a stab at cleanup?